File: original/Code/Combat/comcentergameobj.cpp
#include "comcentergameobj.h"
#include "basecontroller.h"
#include "wwhack.h"
#include "simpledefinitionfactory.h"
#include "persistfactory.h"
#include "definitionmgr.h"
#include "combatchunkid.h"
#include "debug.h"
#include "wwprofile.h"
#include "basecontroller.h"
#include "combat.h"


////////////////////////////////////////////////////////////////
//	Hacks
////////////////////////////////////////////////////////////////
DECLARE_FORCE_LINK (ComCenter)

////////////////////////////////////////////////////////////////
//	Editable and persist factories
////////////////////////////////////////////////////////////////
SimplePersistFactoryClass	<ComCenterGameObjDef,	CHUNKID_GAME_OBJECT_DEF_COMCENTER>						_ComCenterGameObjDefPersistFactory;
SimplePersistFactoryClass	<ComCenterGameObj,		CHUNKID_GAME_OBJECT_COMCENTER>							_ComCenterGameObjPersistFactory;
DECLARE_DEFINITION_FACTORY (ComCenterGameObjDef,	CLASSID_GAME_OBJECT_DEF_COMCENTER, "Com Center")	_ComCenterGameObjDefDefFactory;


////////////////////////////////////////////////////////////////
//	Save/Load constants
////////////////////////////////////////////////////////////////
enum
{
	CHUNKID_DEF_PARENT						=	0x02211153,
	CHUNKID_DEF_VARIABLES,

	MICROCHUNKID_DEF_UNUSED					= 1,
};

enum
{
	CHUNKID_PARENT								=	0x02211154,
	CHUNKID_VARIABLES,

	MICROCHUNKID_UNUSED						= 1,
};


////////////////////////////////////////////////////////////////
//
//	ComCenterGameObjDef
//
////////////////////////////////////////////////////////////////
ComCenterGameObjDef::ComCenterGameObjDef (void)
{
	return ;
}


////////////////////////////////////////////////////////////////
//
//	~ComCenterGameObjDef
//
////////////////////////////////////////////////////////////////
ComCenterGameObjDef::~ComCenterGameObjDef (void)
{
	return ;
}


////////////////////////////////////////////////////////////////
//
//	Get_Class_ID
//
////////////////////////////////////////////////////////////////
uint32
ComCenterGameObjDef::Get_Class_ID (void) const	
{ 
	return CLASSID_GAME_OBJECT_DEF_COMCENTER;
}


////////////////////////////////////////////////////////////////
//
//	Create
//
////////////////////////////////////////////////////////////////
PersistClass *
ComCenterGameObjDef::Create (void) const 
{
	ComCenterGameObj *building = new ComCenterGameObj;
	building->Init (*this);

	return building;
}


////////////////////////////////////////////////////////////////
//
//	Create
//
////////////////////////////////////////////////////////////////
bool
ComCenterGameObjDef::Save (ChunkSaveClass &csave)
{
	csave.Begin_Chunk (CHUNKID_DEF_PARENT);		
		BuildingGameObjDef::Save (csave);
	csave.End_Chunk ();

	csave.Begin_Chunk (CHUNKID_DEF_VARIABLES);
	csave.End_Chunk ();
	return true;
}


////////////////////////////////////////////////////////////////
//
//	Load
//
////////////////////////////////////////////////////////////////
bool
ComCenterGameObjDef::Load (ChunkLoadClass &cload)
{
	while (cload.Open_Chunk ())
	{
		switch (cload.Cur_Chunk_ID ())
		{
			case CHUNKID_DEF_PARENT:
				BuildingGameObjDef::Load (cload);
				break;

			case CHUNKID_DEF_VARIABLES:
				Load_Variables (cload);
				break;
	  
			default:
				Debug_Say (("Unrecognized Com Center Def chunkID\n"));
				break;

		}
		cload.Close_Chunk ();
	}

	return true;
}


////////////////////////////////////////////////////////////////
//
//	Load_Variables
//
////////////////////////////////////////////////////////////////
void
ComCenterGameObjDef::Load_Variables (ChunkLoadClass &cload)
{
	while (cload.Open_Micro_Chunk ()) {
		
		/*switch (cload.Cur_Micro_Chunk_ID ())
		{
			default:
				Debug_Say (("Unrecognized Com Center Def Variable chunkID\n"));
				break;
		}*/

		cload.Close_Micro_Chunk();
	}

	return ;
}


////////////////////////////////////////////////////////////////
//
//	Get_Factory
//
////////////////////////////////////////////////////////////////
const PersistFactoryClass &
ComCenterGameObjDef::Get_Factory (void) const 
{ 
	return _ComCenterGameObjDefPersistFactory; 
}


////////////////////////////////////////////////////////////////
//
//	ComCenterGameObj
//
////////////////////////////////////////////////////////////////
ComCenterGameObj::ComCenterGameObj (void)
{
	return ;
}


////////////////////////////////////////////////////////////////
//
//	~ComCenterGameObj
//
////////////////////////////////////////////////////////////////
ComCenterGameObj::~ComCenterGameObj (void)
{
	return ;
}


////////////////////////////////////////////////////////////////
//
//	Get_Factory
//
////////////////////////////////////////////////////////////////
const PersistFactoryClass &
ComCenterGameObj::Get_Factory (void) const 
{
	return _ComCenterGameObjPersistFactory;
}


////////////////////////////////////////////////////////////////
//
//	Init
//
////////////////////////////////////////////////////////////////
void ComCenterGameObj::Init( void )
{
	Init( Get_Definition() );
}

////////////////////////////////////////////////////////////////
//
//	Init
//
////////////////////////////////////////////////////////////////
void
ComCenterGameObj::Init (const ComCenterGameObjDef &definition)
{
	BuildingGameObj::Init (definition);
	return ;
}


////////////////////////////////////////////////////////////////
//
//	Get_Definition
//
////////////////////////////////////////////////////////////////
const ComCenterGameObjDef &
ComCenterGameObj::Get_Definition (void) const
{
	return (const ComCenterGameObjDef &)BaseGameObj::Get_Definition ();
}


////////////////////////////////////////////////////////////////
//
//	Save
//
////////////////////////////////////////////////////////////////
bool
ComCenterGameObj::Save (ChunkSaveClass &csave)
{
	csave.Begin_Chunk (CHUNKID_PARENT);
		BuildingGameObj::Save (csave);
	csave.End_Chunk ();

	csave.Begin_Chunk (CHUNKID_VARIABLES);
	csave.End_Chunk ();
	return true;
}


////////////////////////////////////////////////////////////////
//
//	Load
//
////////////////////////////////////////////////////////////////
bool
ComCenterGameObj::Load (ChunkLoadClass &cload)
{
	while (cload.Open_Chunk ()) {
		switch (cload.Cur_Chunk_ID ()) {

			case CHUNKID_PARENT:
				BuildingGameObj::Load (cload);
				break;
								
			case CHUNKID_VARIABLES:
				Load_Variables (cload);
				break;

			default:
				Debug_Say (("Unrecognized Com Center chunkID\n"));
				break;
		}

		cload.Close_Chunk();
	}

	return true;
}


////////////////////////////////////////////////////////////////
//
//	Load_Variables
//
////////////////////////////////////////////////////////////////
void
ComCenterGameObj::Load_Variables (ChunkLoadClass &cload)
{
	while (cload.Open_Micro_Chunk ()) {
		
		/*switch (cload.Cur_Micro_Chunk_ID ())
		{
			default:
				Debug_Say (("Unrecognized Com Center Variable chunkID\n"));
				break;
		}*/

		cload.Close_Micro_Chunk();
	}

	return ;
}


////////////////////////////////////////////////////////////////
//
//	On_Destroyed
//
////////////////////////////////////////////////////////////////
void
ComCenterGameObj::On_Destroyed (void)
{
	BuildingGameObj::On_Destroyed ();

	//
	//	Switch off the radar for each player on this team
	//
	if (BaseController != NULL && CombatManager::I_Am_Server ()) {
		BaseController->Check_Radar ();
	}

	return ;
}

File: kotlin-server/server/src/main/kotlin/ccr/server/net/ComCenterGameObj.kt
package ccr.server.net

import ccr.server.defs.ComCenterGameObjDef

// C++: ComCenterGameObj : public BuildingGameObj
// C++ hierarchy: NetworkObject → BaseGameObj → ScriptableGameObj → DamageableGameObj → BuildingGameObj → ComCenterGameObj
class ComCenterGameObj() : BuildingGameObj() {

    // C++: ComCenterGameObj() — empty constructor body; no additional fields

    init {
        // C++: no constructor body beyond default initialization
    }

    // Secondary constructor for tests — chains to BuildingGameObj secondary constructor.
    constructor(
        definitionId: Int,
        position: ccr.math.Vector3 = ccr.math.Vector3(),
        sphereCenter: ccr.math.Vector3 = ccr.math.Vector3(),
        sphereRadius: Float = 50f,
        health: Float = 0f,
        isDestroyed: Boolean = false,
        isPowerOn: Boolean = true,
    ) : this() {
        definition = ccr.server.defs.BaseGameObjDef(
            name = "comcenter_$definitionId", id = definitionId.toUInt(), chunkId = 0u
        )
        this.position = position
        this.collectionSphere = ccr.server.level.Sphere(sphereCenter, sphereRadius)
        this.isDestroyed = isDestroyed
        this.isPowerOn = isPowerOn
        defenseObject.health = health
    }

    // C++: ~ComCenterGameObj() — empty destructor body
    override fun destruct() {
        super.destruct()
    }

    // C++: virtual void Init(void)
    override fun init() {
        init(getComCenterDefinition())
    }

    // C++: void Init(const ComCenterGameObjDef& definition)
    fun init(definition: ComCenterGameObjDef) {
        super.init(definition)
    }

    // C++: const ComCenterGameObjDef& Get_Definition() const
    fun getComCenterDefinition(): ComCenterGameObjDef = definition as ComCenterGameObjDef

    // C++: ComCenterGameObj* As_ComCenterGameObj() { return this; }
    // → handled by Kotlin type system: `this is ComCenterGameObj`

    // C++: virtual bool Save(ChunkSaveClass& csave)
    override fun save(csave: ChunkSaveClass): Boolean {
        csave.beginChunk(CHUNKID_PARENT)
        super.save(csave)
        csave.endChunk()

        csave.beginChunk(CHUNKID_VARIABLES)
        csave.endChunk()

        return true
    }

    // C++: virtual bool Load(ChunkLoadClass& cload)
    override fun load(cload: ChunkLoadClass): Boolean {
        while (cload.openChunk()) {
            when (cload.curChunkId) {
                CHUNKID_PARENT    -> super.load(cload)
                CHUNKID_VARIABLES -> loadVariables(cload)
                else -> error("Unrecognized ComCenterGameObj chunk ID: ${cload.curChunkId}")
            }
            cload.closeChunk()
        }
        return true
    }

    // C++: void Load_Variables(ChunkLoadClass& cload)
    private fun loadVariables(cload: ChunkLoadClass) {
        while (cload.openMicroChunk()) {
            when (cload.curMicroChunkId) {
                MICROCHUNKID_UNUSED -> { /* no-op: legacy unused micro-chunk */ }
                else -> error("Unrecognized ComCenterGameObj variable chunk ID: ${cload.curMicroChunkId}")
            }
            cload.closeMicroChunk()
        }
    }

    // C++: void On_Destroyed()
    override fun onDestroyed() {
        super.onDestroyed()

        // C++: if (BaseController != NULL && CombatManager::I_Am_Server()) BaseController->Check_Radar()
        // (always server — omit I_Am_Server guard per guide)
        baseController?.checkRadar()
    }

    companion object {
        // C++: comcentergameobj.cpp enum { CHUNKID_PARENT = 0x02211154, CHUNKID_VARIABLES, ... }
        private const val CHUNKID_PARENT    = 0x02211154
        private const val CHUNKID_VARIABLES = 0x02211155

        private const val MICROCHUNKID_UNUSED = 1
    }
}
